src/openforms/formio/api/serializers.py

```diff
@@ -44,3 +46,24 @@ def get_url(self, instance) -> str:
             kwargs={"uuid": instance.uuid},
             request=request,
         )
+
+    def validate_file(self, value: UploadedFile):
+        if value.content_type == "application/octet-stream":
+            ext = value.name.split(".")[-1]
+            m = magic.Magic(extension=True)
+            extensions = m.from_buffer(value.read(2048)).split("/")
+            if extensions == ["???"]:
+                pass  # magic db doesn't know extensions; let form validation decide on mime-type
+            elif ext not in extensions:
+                raise serializers.ValidationError(
+                    _("The file %(filename)s is not a %(file-type)s.")
+                    % {"filename": value.name, "file-type": f".{ext}"}
+                )
+        elif magic.from_buffer(value.read(2048), mime=True) != value.content_type:
+            raise serializers.ValidationError(
+                _("The file %(filename)s is not a %(file-type)s.")
+                % {"filename": value.name, "file-type": value.content_type}
+            )
+
+        # valid
+        return value
```

Comment:
Can we implement this as a re-usable DRF validator class/function, and then call that as part of the `openforms.submissions.attachments.validate_uploads` flow? I'd like to keep this validation machinery/logic in a single place to avoid implementations diverging in logic.

Reply:
I've refactored for re-use. I didn't squash the last refactor commit, because the :beers: gitmoji could be equally valid. Co-author Hertog Jan :wink:. Have a nice weekend.